torch/_decomp/decompositions_for_rng.py

```diff
@@ -0,0 +1,292 @@
+    relative_offset. seed and base_offset basically point to the rng state just
+    before tracing starts. relative offset tracks the totaly consumed offset at
+    trace time.
+    """
+
+    def __init__(self):
+        self.reset()
+
+    def reset(self):
+        self.seed = torch.Tensor(0)
+        self.base_offset = torch.Tensor(0)
+        self.relative_offset = 0
+
+    def advance_offset(self, consumed_offset):
+        self.relative_offset += consumed_offset
+
+    def set_state(self, seed, base_offset, relative_offset=0):
+        self.seed = seed
+        self.base_offset = base_offset
+        self.relative_offset = relative_offset
+
+    def get_state_as_tuple(self):
+        return (self.seed, self.base_offset + self.relative_offset)
+
+    def get_state_as_tensor(self):
+        # Only needed because we override get_rng_state.
+        seed_portion = self.seed.reshape(1)
+        offset_portion = (self.base_offset + self.relative_offset).reshape(1)
+        return torch.cat([seed_portion, offset_portion])
+
+    def set_state_from_tensor(self, state):
+        # Only needed because we override set_rng_state.
+        seed, offset = torch.split(state, 1)
+        self.seed = seed[0]
+        self.base_offset = offset[0]
+        self.relative_offset = 0
+
+
+class PhiloxTotalOffsets:
+    """
+    PhiloxStateTracker computes the total fwd and bwd offsets for an AOT
+    Autograd traced graph. However, PhiloxStateTracker is a singleton class, but
+    the total offsets are specific to each traced graph. These offsets are
+    stored as part of AOT graph. This class just encapsulates the fwd and bwd
+    offsets to be used at runtime.
+    """
+
+    def __init__(self, fwd_offset=0, bwd_offset=0):
+        self.total_fwd_offset = fwd_offset
+        self.total_bwd_offset = bwd_offset
+
+
+class PhiloxStateTracker:
+    """
+    Singleton class to track the philox rng state during AOT Autograd tracing.
+    For each aot tracing instance, AOT Autograd resets this tracker and keeps
+    track of both forward and backward offsets. At runtime, we only care about
+    the total consumed forward and backward offsets. There are stored as part of
+    aot_config (PhiloxTotalOffsetsForRuntime), which is a config object per
+    aot-tracing.
+    """
+
+    running_state = PhiloxState()
+    fwd_state = PhiloxState()
+    bwd_state = PhiloxState()
+
+    @classmethod
+    def reset(cls):
+        cls.running_state = PhiloxState()
+        cls.fwd_state = PhiloxState()
+        cls.bwd_state = PhiloxState()
+
+    @classmethod
+    def mark_beginning_of_forward(cls):
+        # Tells the tracker to use fwd_state as the running state
+        cls.running_state = cls.fwd_state
+
+    @classmethod
+    def mark_beginning_of_backward(cls):
+        # Tells the tracker to use bwd_state as the running state
+        cls.running_state = cls.bwd_state
+
+    @classmethod
+    def record_state(cls, seed, offset, mode):
```

Comment:
We use this to setup the initial fwd and bwd state tensors. `running_state` basically points to either `fwd_state` or `bwd_state`. However, this functions just sets the initial state for `fwd` and `bwd`, so that AOTDispatch can track the ProxyTensors while tracing the joint graph, and make right connections.

Reply:
Why not just make sure the current mode is already set before we call this?